File: clients/rust/src/instructions/manage_controller.rs
use crate::{
    derive_controller_authority_pda, derive_permission_pda,
    generated::{instructions::ManageControllerBuilder, types::ControllerStatus},
};
use solana_instruction::Instruction;
use solana_pubkey::Pubkey;


pub fn create_manage_controller_instruction(
    controller: &Pubkey,
    authority: &Pubkey,
    status: ControllerStatus,
) -> Instruction {
    let calling_permission_pda = derive_permission_pda(controller, &authority);
    let controller_authority = derive_controller_authority_pda(controller);

    ManageControllerBuilder::new()
        .status(status)
        .controller(*controller)
        .controller_authority(controller_authority)
        .authority(*authority)
        .permission(calling_permission_pda)
        .program_id(crate::SVM_ALM_CONTROLLER_ID)
        .instruction()
}

File: clients/ts/src/instructions/manageController.ts
import { Address, TransactionSigner } from "@solana/kit";
import { getManageControllerInstruction } from "../generated/instructions/manageController";
import {
  deriveControllerAuthorityPda,
  derivePermissionPda,
} from "../pdas";
import { SVM_ALM_CONTROLLER_PROGRAM_ADDRESS } from "../generated";
import type { ControllerStatusArgs } from "../generated/types";

/**
 * Instruction generation for managing a controller account
 */
export async function createManageControllerInstruction(
  controller: Address,
  authority: TransactionSigner,
  status: ControllerStatusArgs
) {
  const permissionPda = await derivePermissionPda(controller, authority.address);
  const controllerAuthority = await deriveControllerAuthorityPda(controller);

  return getManageControllerInstruction({
    controller,
    controllerAuthority,
    authority,
    permission: permissionPda,
    programId: SVM_ALM_CONTROLLER_PROGRAM_ADDRESS,
    status,
  });
}

